include/pybind11/iostream.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include "pybind11.h"
+
+#include <streambuf>
+#include <ostream>
+#include <string>
+#include <memory>
+
+NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+NAMESPACE_BEGIN(detail)
+
+class pythonbuf : public std::streambuf {
+private:
+    typedef std::streambuf::traits_type traits_type;
+    char          d_buffer[1024];
+
+    object pyostream;
+
+    int overflow(int c) {
+        if (!traits_type::eq_int_type(c, traits_type::eof())) {
+            *this->pptr() = traits_type::to_char_type(c);
+            this->pbump(1);
+        }
+        return this->sync() ? traits_type::not_eof(c) : traits_type::eof();
+    }
+
+    int sync() {
+        if (this->pbase() != this->pptr()) {
+            std::string line(this->pbase(), this->pptr());
+
+            auto write = pyostream.attr("write");
+            write(line);
+            pyostream.attr("flush")();
+
+            this->setp(this->pbase(), this->epptr());
+        }
+        return 0;
+    }
+public:
+    pythonbuf(object pyostream)
+        : pyostream(pyostream) {
+        this->setp(this->d_buffer, this->d_buffer + sizeof(this->d_buffer) - 1);
+    }
+};
+
+class opythonstream : private virtual pythonbuf, public std::ostream {
```

Comment:
Not trivially, since it will be constructed in the wrong order, as the final object is required to construct the class it would be a member of. Let me know if you have any ideas other than manually building objects first and then passing around smart pointers or similar.